File: haskell/test/EmissionSource.hs
{-# LANGUAGE UndecidableInstances #-}

module EmissionSource where

import Control.Applicative ((<|>))
import Data.Function ((&))
import Data.Map (Map)
import Data.Map qualified as Map
import Lang.TSHelper

calculateEmissionsInTons ::
  Maybe JSNumber ->
  Maybe JSNumber ->
  EmitterTable ->
  Maybe EmitterID ->
  JSNumber
calculateEmissionsInTons quantity factor factors emitter_id =
  if (quantity /= Nothing && factor /= Nothing)
    then (quantity !) * (factor !) / 1000
    else
      let table_factor = getFactorFromDatabase factors emitter_id
       in if quantity /= Nothing && table_factor /= Nothing
            then (quantity !) * (table_factor !) / 1000
            else 0

getFactorFromDatabase :: EmitterTable -> Maybe EmitterID -> Maybe JSNumber
getFactorFromDatabase results = (>>= (`Map.lookup` results))

type EmitterID = String

type JSNumber = Double

data ManuelOrDatabase = Manual | Database
  deriving (Show)

type EmitterTable = Map EmitterID JSNumber

data EmitterOptions = EmitterOptions
  { assignee :: Maybe String,
    quantity :: Maybe ValueWithUncertainty,
    factor :: Maybe ValueWithUncertainty,
    emissions :: Maybe ValueWithUncertainty,
    emitterId :: Maybe EmitterID
  }

data ValueWithUncertainty = ValueWithUncertainty
  { value :: JSNumber,
    uncertainty :: JSNumber
  }
  deriving (Eq)

data Emitter = Emitter
  { description :: String,
    getEmissions :: EmitterTable -> JSNumber,
    getUncertaintyOfEmissions :: JSNumber,
    options :: EmitterOptions
  }

data EmitterFormValues f = EmitterFormValues
  { description :: f String,
    assignee :: Maybe String,
    category :: f String,
    quantity :: Maybe String,
    quantityUncertainty :: Maybe JSNumber,
    factor :: Maybe String,
    factorUncertainty :: Maybe JSNumber,
    factorMode :: f ManuelOrDatabase,
    enterEmissionsManually :: f Bool,
    emissions :: Maybe String,
    emissionsUncertainty :: Maybe JSNumber
  }

deriving instance
  (Show (f String), Show (f ManuelOrDatabase), Show (f Bool)) =>
  Show (EmitterFormValues f)

-- >>> getInitialFormValues (Emitter "a" (\_ -> 9.0) 3.2 (EmitterOptions Nothing Nothing Nothing Nothing Nothing)) mempty
-- EmitterFormValues {description = Just "a", assignee = Nothing, category = Nothing, quantity = Nothing, quantityUncertainty = Just 30.0, factor = Nothing, factorUncertainty = Just 30.0, factorMode = Just Database, enterEmissionsManually = Just False, emissions = Just "0.0", emissionsUncertainty = Just 30.0}
getInitialFormValues :: Emitter -> EmitterTable -> EmitterFormValues Maybe
getInitialFormValues emitter factors =
  let valuesFromEmitter =
        EmitterFormValues
          { description = Just emitter.description,
            assignee = emitter.options.assignee,
            quantity = emitter.options.quantity ? (.value) ? show,
            quantityUncertainty = emitter.options.quantity ? (.uncertainty) ?? 30 & Just,
            emissionsUncertainty = emitter.options.emissions ? (.uncertainty) ?? 30 & Just,
            factorUncertainty = emitter.options.factor ? (.uncertainty) ?? 30 & Just,
            category = Nothing,
            factor = Nothing,
            factorMode = Nothing,
            enterEmissionsManually = Nothing,
            emissions = Nothing
          }
   in if emitter.options.emissions /= Nothing
        then
          valuesFromEmitter
            { enterEmissionsManually = Just True,
              emissions = emitter.options.emissions ? (.value) ? show
            }
        else
          valuesFromEmitter
            { factorMode =
                Just $
                  if emitter.options.factor == Nothing then Database else Manual,
              factor = (emitter.options.factor ? (.value) <|> getFactorFromDatabase factors emitter.options.emitterId) ? show,
              enterEmissionsManually = Just False,
              emissions =
                calculateEmissionsInTons
                  (emitter.options.quantity ? (.value))
                  (emitter.options.factor ? (.value))
                  factors
                  emitter.options.emitterId
                  & show
                  & Just
            }

File: sum-types/typescript/sum-types/src/emission-source.ts
export type ValueWithUncertainty = Readonly<{
  value: number;
  uncertainty: number;
}>;

export const EMISSION_FACTOR_TABLE = {
  ELECTRICITY_GERMANY: 0.4,
  NATURAL_GAS: 0.2,
} as const;

type EmitterFactors = typeof EMISSION_FACTOR_TABLE;

function getFactorFromDatabase(
  factors: EmitterFactors,
  emitterId: string,
): number | undefined {
  return factors[emitterId as keyof EmitterFactors];
}

export function calculateEmissionsInTons(
  quantity?: number,
  factor?: number,
  factorsFromDatabase?: EmitterFactors,
  emitterId?: string,
): number {
  if (quantity && factor) {
    return (quantity * factor) / 1000;
  } else if (
    quantity &&
    factorsFromDatabase &&
    emitterId &&
    getFactorFromDatabase(factorsFromDatabase, emitterId)
  ) {
    return (
      (quantity * getFactorFromDatabase(factorsFromDatabase, emitterId)!) / 1000
    );
  } else {
    return 0;
  }
}

export type Emitter = {
  description: string;
  emitterId?: string;
  quantity?: ValueWithUncertainty;
  factor?: ValueWithUncertainty;
  emissions?: ValueWithUncertainty;
  getEmissions: (emitterFactors: EmitterFactors) => number;
  getUncertaintyOfEmissions: () => number;
};

export function makeEmitter(
  options: Omit<Emitter, "getEmissions" | "getUncertaintyOfEmissions">,
): Emitter {
  function getFactor(emitterFactors: EmitterFactors): number | undefined {
    if (Number.isFinite(options?.factor?.value)) {
      return options?.factor?.value;
    } else if (options?.emitterId) {
      return emitterFactors[options?.emitterId]?.factor;
    } else {
      return undefined;
    }
  }

  return {
    ...options,
    getEmissions(emitterFactors) {
      if (options?.emissions != null) {
        return options.emissions.value;
      }
      if (options?.quantity == null) {
        return 0;
      }

      const factor = getFactor(emitterFactors);
      if (factor == null) {
        return 0;
      }

      return (factor * options.quantity.value) / 1000;
    },
    getUncertaintyOfEmissions() {
      if (options?.emissions != null) {
        return options.emissions.uncertainty;
      }

      if (options?.quantity != null && options?.factor != null) {
        return Math.sqrt(
          Math.pow(Number(options?.quantity?.uncertainty), 2) +
            Math.pow(Number(options?.factor?.uncertainty), 2),
        );
      }

      return 60;
    },
  };
}

export type EmitterFormValues = {
  description: string;
  quantity?: string;
  quantityUncertainty?: number;
  emitterId?: keyof EmitterFactors;
  factor?: string;
  factorUncertainty?: number;
  factorMode: "manual" | "database";
  enterEmissionsManually: boolean;
  emissions?: string;
  emissionsUncertainty?: number;
};

export function getInitialFormValues(
  emitter: Emitter,
  factors: EmitterFactors,
): Partial<EmitterFormValues> {

  let valuesFromEmitter: Partial<EmitterFormValues> = {
    description: emitter.description,
    emitterId: Object.keys(factors).includes(emitter.emitterId) ? emitter.emitterId as keyof EmitterFactors : undefined,
    quantity: emitter.quantity?.value.toString(),
    quantityUncertainty: emitter.quantity?.uncertainty ?? 30,
    emissionsUncertainty: emitter.emissions?.uncertainty ?? 30,
    factorUncertainty: emitter.factor?.uncertainty ?? 30,
  };
  if (emitter.emissions != null) {
    // manual mode, emissions are entered directly
    valuesFromEmitter = {
      ...valuesFromEmitter,
      enterEmissionsManually: true,
      emissions: emitter.emissions?.value.toString(),
    };
  } else {
    valuesFromEmitter = {
      ...valuesFromEmitter,
      factorMode: emitter.factor == null ? "database" : "manual",
      factor:
        emitter.factor?.value.toString() ??
        getFactorFromDatabase(factors, emitter.emitterId)?.toString(),
      enterEmissionsManually: false,
      emissions: calculateEmissionsInTons(
        emitter.quantity?.value,
        emitter.factor?.value,
        factors,
        emitter.emitterId,
      ).toString(),
    };
  }

  return valuesFromEmitter;
}
